File: old/2016/09_explosives_in_cyberspace/aoc09.py
"""Explosives in Cyberspace

Advent of Code 2016, day 9
Solution by Geir Arne Hjelle, 2016-12-09
"""
# Standard library imports
import itertools
import pathlib
import sys


def find_length(string, iterate):
    s_iter = iter(string)
    for c in s_iter:
        if c != "(":
            yield 1
            continue

        counter, repeats = [
            int(n)
            for n in "".join(itertools.takewhile(lambda c: c != ")", s_iter)).split("x")
        ]
        data = itertools.islice(s_iter, counter)
        yield repeats * (sum(find_length(data, True)) if iterate else len(list(data)))


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    with open(file_path, mode="r") as fid:
        for line in fid:
            string = line.strip()
            print(f"{string[:40]:<40s} - Method one: {sum(find_length(string, False))}")
            print(f"{string[:40]:<40s} - Method two: {sum(find_length(string, True))}")


if __name__ == "__main__":
    main(sys.argv[1:])

File: old/2020/09_encoding_error/aoc09.jl
# Encoding Error
#
# Advent of Code 2020, day 9
# Solution by Geir Arne Hjelle, 2020-12-09

module AOC09

using Pipe


# Find two numbers that add up to target. Will return all numbers that are part
# of such pairs.
function find_adders(numbers, target)
    lookup = Set(numbers)
    [n for n in numbers if target - n in lookup]
end


# Find 2 or more contiguous numbers adding up to target. Work backwards to
# avoid working through long runs of small numbers
function find_contiguous(numbers, target)
    idx = length(numbers)
    while idx > 2
        for run_length in 2:idx
            run_sum = sum(numbers[idx - run_length:idx])
            if run_sum > target
                break
            elseif run_sum == target
                return numbers[idx - run_length:idx]
            end
        end
        idx -= 1
    end
end

# Check run of numbers for first error in encoding
function check_numbers(numbers, preamble)
    for idx in preamble + 1:length(numbers)
        if isempty(find_adders(numbers[idx - preamble:idx], numbers[idx]))
            return numbers[idx]
        end
    end
end


# Solve the problem for one file
function solve(input)
    # Parse input, length of preamble has been manually added on first line
    input = split(input, "\n") .|> ln -> parse(Int64, ln)
    preamble, numbers = input[1], input[2:end]

    # Part 1
    part_1 = check_numbers(numbers, preamble)

    # Part 2
    run = find_contiguous(numbers, part_1)
    part_2 = run |> extrema |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
